clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -381,23 +441,37 @@ public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
                 requestTopicDataMap.put(topicPartition.topic(), topic);
             }
 
-            OffsetCommitRequest.Builder builder = new OffsetCommitRequest.Builder(
-                new OffsetCommitRequestData()
+            OffsetCommitRequestData data = new OffsetCommitRequestData()
                     .setGroupId(this.groupId)
-                    .setGenerationIdOrMemberEpoch(generation.generationId)
-                    .setMemberId(generation.memberId)
-                    .setGroupInstanceId(groupInstanceId)
-                    .setTopics(new ArrayList<>(requestTopicDataMap.values())));
+                    .setGroupInstanceId(groupInstanceId.orElse(null))
+                    .setTopics(new ArrayList<>(requestTopicDataMap.values()));
+            if (memberInfo.memberId.isPresent()) {
+                data = data.setMemberId(memberInfo.memberId.get());
+            }
+            if (memberInfo.memberEpoch.isPresent()) {
+                data = data.setGenerationIdOrMemberEpoch(memberInfo.memberEpoch.get());
+            }
+
+            OffsetCommitRequest.Builder builder = new OffsetCommitRequest.Builder(data);
+
             NetworkClientDelegate.UnsentRequest resp = new NetworkClientDelegate.UnsentRequest(
                 builder,
                 coordinatorRequestManager.coordinator());
             resp.whenComplete(
                 (response, throwable) -> {
                     try {
                         if (throwable == null) {
+                            log.debug("OffsetCommit response received for offsets {} ", offsets);
                             onResponse(response);
                         } else {
-                            onError(throwable, resp.handler().completionTimeMs());
+                            log.debug("OffsetCommit completed with error for offsets {}", offsets, throwable);
+                            long currentTimeMs = resp.handler().completionTimeMs();
+                            handleCoordinatorDisconnect(throwable, currentTimeMs);
+                            if (throwable instanceof RetriableException) {
+                                retry(currentTimeMs, throwable);
+                            } else {
+                                future.completeExceptionally(throwable);
+                            }
```

Comment:
I wonder if we could have something like:

```
if (notRetriable(throwable) || !maybeRetry(currentTimeMs, throwable)) {
    future.completeExceptionally(throwable);
}
```

Thinking about this because `retry` may not retry if the timer is expired.

Reply:
The `retry` already fails the request future if it does not retry (ex. timer expired). That's needed from all the places where we call the retry, so seemed sensible to keep it centralized there.